<a href="https://colab.research.google.com/github/rii92/Datmin/blob/main/001_the_basics.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Objective:

Can understanding that programming is basically manipulating data to fit into a specific requirement.

# What is programming?

![image.png](https://storage.googleapis.com/rg-ai-bootcamp/python-prep/programming-main-focus-min.png)

The real foundation of programming, the essence of it is to manipulate data to adhere to what we need to do next.

This "what we need to do next" itself is depends on what kind of programming we currently doing. For example, if we're doing a web programming, the "thing we need to do next after manipulating data" might be to render a web page, saving user data to database, or sending an email. If we're doing a game programming, what we need to do might be to render a game screen, playing a sound, or saving the game state.

Let's check below simple program:

In [1]:
#@title Preparation so we can do get_emotion(data), you can just run and ignore this line of code for now
%pip install transformers
%pip install torch
from transformers import pipeline

classifier = pipeline(task="text-classification", model="SamLowe/roberta-base-go_emotions", top_k=None)

def get_emotion(text):
    model_outputs = classifier(text)
    return highest_confidence_label(model_outputs[0])

def highest_confidence_label(data):
    highest_score = max(data, key=lambda x:x['score'])
    confidence = round(highest_score['score'] * 100, 2)
    return f"{highest_score['label']}, {confidence}% confidence"

  Using cached nvidia_cuda_nvrtc_cu12-12.1.105-py3-none-manylinux1_x86_64.whl (23.7 MB)
  Using cached nvidia_cuda_runtime_cu12-12.1.105-py3-none-manylinux1_x86_64.whl (823 kB)
  Using cached nvidia_cuda_cupti_cu12-12.1.105-py3-none-manylinux1_x86_64.whl (14.1 MB)
  Using cached nvidia_cudnn_cu12-8.9.2.26-py3-none-manylinux1_x86_64.whl (731.7 MB)
  Using cached nvidia_cublas_cu12-12.1.3.1-py3-none-manylinux1_x86_64.whl (410.6 MB)
  Using cached nvidia_cufft_cu12-11.0.2.54-py3-none-manylinux1_x86_64.whl (121.6 MB)
  Using cached nvidia_curand_cu12-10.3.2.106-py3-none-manylinux1_x86_64.whl (56.5 MB)
  Using cached nvidia_cusolver_cu12-11.4.5.107-py3-none-manylinux1_x86_64.whl (124.2 MB)
  Using cached nvidia_cusparse_cu12-12.1.0.106-py3-none-manylinux1_x86_64.whl (196.0 MB)
  Using cached nvidia_nccl_cu12-2.20.5-py3-none-manylinux2014_x86_64.whl (176.2 MB)
  Using cached nvidia_nvtx_cu12-12.1.105-py3-none-manylinux1_x86_64.whl (99 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 21.3/21

/usr/local/lib/python3.10/dist-packages/huggingface_hub/utils/_token.py:89: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


config.json:   0%|          | 0.00/1.92k [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/499M [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/380 [00:00<?, ?B/s]

vocab.json:   0%|          | 0.00/798k [00:00<?, ?B/s]

merges.txt:   0%|          | 0.00/456k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/2.11M [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/280 [00:00<?, ?B/s]

In [2]:
get_emotion("I'm so excited to meet you guys there!")

'excitement, 81.15% confidence'

In above line of code we can see that:

- We have a data that we want to process: "I'm so excited to meet you guys there!"
- We have a requirement that we want to do: Get the emotion probability of the data using get_emotion("we put our data here")
- And it return a response of the confidence of the emotion probability: excitement, 81.15% confidence

So we have an input, we process it, and we get an output, we can use that output as an input again, and so on. That's the essence of programming.

> Please have fun: While you're at it, have fun change the data above (change inside the `"` symbol) to feed it with your own data, and see what kind of emotion probability that you get. Try to change it something like "Why are you running late?", "Stuck in traffic, ugh!", "I'm not feeling well today, sorry guys"

# Challenge

## Challenge 1

For the first challenge we'll start with a really simple task:

Change the `get_emotion("I'm so excited to meet you guys there!")` to "I'm really stressed out for this meeting, I hope it goes well" and send the result of the probability.

In [3]:
get_emotion("I'm really stressed out for this meeting, I hope it goes well") #Change the text inside get_emotion()

'optimism, 39.45% confidence'

To send the result:
1. Input your student_id and name in the box below
2. Change the text "my result" below with the text output from get_emotion() above. example: "excitement, 81.15% confidence"
3. Run the code block by pressing the play button in the menu above.

In [4]:
!pip install rggrader

from rggrader import submit

# @title #### Student Identity
student_id = "REAFCDNE" # @param {type:"string"}
name = "Riofebri Prasetia" # @param {type:"string"}

result = "optimism, 39.45% confidence"

# Submit Method
assignment_id = "001_the-basics"
question_id = "01_get_emotion"
submit(student_id, name, assignment_id, result, question_id)

'Assignment successfully submitted'

# Uh oh, the input changed!

As we said above, one of the most fundamental thing for us to understand programming is to manipulate data. For above code, we have a convenient text that we can just pass to `get_emotion`, but what if the data isn't that convenient? Such as we have two inputs like so:

```
"I'm so {emotion} to meet you guys there!"
"excited"
```

We have two separate inputs, and, to make sure that we can process that input to `get_emotion`, we need to make sure that input is compatible with `get_emotion` requirement. Which can be done like so:

In [5]:
("I'm so {emotion} to meet you guys there!".replace("{emotion}", "happy"))

"I'm so happy to meet you guys there!"

In [6]:
get_emotion("I'm so {emotion} to meet you guys there!".replace("{emotion}", "happy"))

'joy, 88.13% confidence'

In [ ]:
get_emotion("I'm so {emotion} to meet you guys there!".replace("emotion", "happy")) #Change the text inside get_emotion()
print("I'm so {emotion} to meet you guys there!".replace("{emotion}", "happy"))

So now we're talking! If we see above, we'll see additional code after our "I'm so {emotion} to meet you guys there!" data, which is `.replace("{emotion}", "excited")`.

We'll learn about how that works later, but for now, let's just say that we're replacing the `{emotion}` part of the data with the "excited" data, as the `get_emotion` function requires to have only a single text data as an input.

# Conclusion

As we said above, the complexity on programming mostly rely only on making sure that your data is ready to be inputted to our given task. When we want to save our data to database, we want to manipulate our data so it can be saved properly. If we want to render a web page, we want to manipulate our data so it can be rendered properly. And so on.

It's simple right?

# Terminology alert!

Let's return to our first line of code:
    
```
get_emotion("I'm so excited to meet you guys there!")
```

- `get_emotion` is a function
- `"I'm so excited to meet you guys there!"` is a data with a data type of string

We'll learn more about function and data type later, and other terminology that's related to function and data type, but for now just remember above terminology.